### DISCUSSION TITLE
Nautobot Modules

### DISCUSSION THREAD

**Discussion by jvanderaa** [ansible-collections/ansible-inclusion#28] [Second review in progress]
Looking to add a collection to the Ansible Community installation. 

The Nautobot collection is used to create, update, and delete Nautobot objects. Nautobot is an Open Source source of truth and network automation platform. These modules are used at the core of workflows to manage objects within the Nautobot project. These modules were forked from the netbox.netbox collection earlier in 2021 and have additional modules added such as a GraphQL action plugin + lookup plugin.

Ansible Galaxy: https://galaxy.ansible.com/networktocode/nautobot
GitHub: https://github.com/nautobot/nautobot-ansible/
Issues Tracker: https://github.com/nautobot/nautobot-ansible/issues
GitHub Handles: @networktocode
Is the collection part of Automation Hub: No
We meet [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst): Yes

**Comment by tadeboro** [comment]
Thank you for your submission, @jvanderaa.

The items in the checklist can be in one of the following states:

  * Items with a checkmark and no attached comment are OK and need no intervention from your side.
  * Items with a checkmark that contain a comment will not block the inclusion process, but by following the advice given in the comment, the collection's quality can be improved.
  * Items with no checkmark and no extra details are waiting for evaluation.
  * Items with no checkmark that contain a comment need your attention.

Feel free to ask any questions and provide feedback. Thanks in advance for your understanding and patience.

_For details about the following points, refer to the [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst)._

Every comment should say whether the reviewer expects it to be addressed, or whether it's optional.

**Public availability and communication:**
- [x] published on Ansible Galaxy [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [x] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
- [x] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
      **COMMENT:** The CoC [passed the approval](https://meetbot.fedoraproject.org/ansible-diversity/2021-09-16/ansible_diversity_working_group_%7C_agenda:_https:github.comansiblecommunityissues577.2021-09-16-19.05.html). The [full meeting log](https://meetbot.fedoraproject.org/ansible-diversity/2021-09-16/ansible_diversity_working_group_|_agenda:_https:github.comansiblecommunityissues577.2021-09-16-19.05.log.html) contains some information on what could be improved, but there is nothing in the CoC that would block the inclusion.
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues
- [x] have a public SCM repository, releases of the collection are tagged in this repository

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
- [x] follows [licensing rules](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#licensing)
- [x] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
- [x] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html); as well as these other requirements:
  - [x] modules that only gather information are named `<something>_info`
  - [x] modules that return `ansible_facts` are named `<something>_facts` and do not return non-facts
  - [x] other modules must not allow querying information using specific `state` option values, or similar mechanisms (like `state=get` or `state=query`).  These features should be moved to `<something>_info` or `<something>_fact` modules.
  - [x] `check_mode` is supported in all `_info` and `_facts` modules
- [x] supports Python 2.6 or greater and Python 3.5 or greater. If it does not, read the [full guidelines](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#python-compatibility) to see if you qualify for an exception and document the unsupported [Python versions](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#ansible-and-python-3) in the collection ``README.md`` and in every module and plugin (or in doc fragments)
      **COMMENT:** Modules only support Python >= 3.6 because the underlying Python client supports those Python versions.
- [x] only uses the [allowed plugin types](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#modules-plugins) in the `plugins/` directory
- [x] has `README.md`
- [x] documentation, examples, and return sections use FQCNs for the `M(..)` [format macros](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation) when referring to modules
- [x] modules (or plugins) from ansible-core use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module or plugin was added
- [x] FQCNs are used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-core

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0, and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
- [x] has changelog, preferably with `changelogs/changelog.yaml`

**Tests:**
- [x] passed `ansible-test sanity`
      **SHOULD FIX:** The `ansible-test sanity --docker`  command is still not clean because the ignore files are missing.
- [x] if `test/sanity/ignore*.txt` exists, it does not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [x] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
      **COMMENT:** The CI setup is a bit disorganized, but sanity tests do indeed run on all supported Ansible versions as part of the integration test job.
- [x] all CI tests run against every pull request
- [x] all CI tests run against a commit that releases the collection; if they don't pass, the collection won't be released
- [x] all CI tests run regularly (nightly, or at least once per week)

**Reply by tadeboro** [reply]
I did not finish my review since there seem to be some major issues with the licensing in this collection. For example, commit 301e66c8 stripped copyright and licensing information from modules (https://github.com/nautobot/nautobot-ansible/commit/301e66c846fda9641b15aa50a96c704f53bad681#diff-1aaebd78d9857e913ef7e74ca7a6448ee8bf3a04a583666f7882e56b378e98a4 is one such module modification). As far as I can tell, this is not legal since it violates the original copyright and license because the new repository seems to be under Apache 2.0 now (again, I am not sure if changing the license here was a valid move).

/cc @abadger @felixfontein @gundalow

**Reply by gundalow** [reply]
I've spoken with Networktocode, then are talking internally about the license change.
They are aware that to relicense the code would require approval of every person who's ever contributed.

**Reply by jvanderaa** [reply]
@tadeboro the licensing should be squared away on the repo. I just wanted to double check that the two MUST FIXs are the only items outstanding then? I am working on getting those incorporated this week.

**Reply by tadeboro** [reply]
No, I stopped reviewing when I hit the licensing issues. I did not look into the unchecked items without any comments. I will try to look at them as soon as possible, but this may take a week or two because AnsibleFest session preparation and recording is in full swing right now.

**Reply by jvanderaa** [reply]
Got it, so what was posted above is the checklist to go through, that I can start to take a look at the items ahead of time.

**Reply by tadeboro** [reply]
I did a full review of the collection and it is still not ready for inclusion in Ansible 5.

@jvanderaa As for looking at the items ahead of time: when you submitted the collection for inclusion, you said the collection meets the criteria for inclusion. So I would expect that most of the items on our checklist should be OK by now.

**Reply by jvanderaa** [reply]
We'll get on those this week yet. 

As it relates to the license on the individual files. Is it not that if there is not a specific license on the file itself, that it would then go back to the license within the repository? Each individual file should not have to have a license indicator on them if I understand things properly.

**Reply by jvanderaa** [reply]
@tadeboro everything here should be addressed at this time.

## Releasing Updates

The verbiage has been updated.  

The 2.0 release is still valid if not on ansible-core. WIth ansible-core there became an incompatibility that required us to release a new version. Per SemVer this was a breaking change, thus the need to move forward. We did not have a method to move this back. A notification was also posted in discussions that mentioned the upcoming breaking change - https://github.com/nautobot/nautobot-ansible/discussions/70

## CoC Updates

This was a clone from the Ansible CoC. If there needs to be updates, we will gladly updated them. The only difference was the markdown lint that a bolded section should be a header, not a bolded word. So that is what was changed that deviated from the Ansible CoC - [https://docs.ansible.com/ansible/latest/community/code_of_conduct.html](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html).

## Licensing Headers

The files had license headers added rather than implying that the project licensing was to be used.

## Parameters Version Added

`version_added` was added to all of the modules that had the change, which was, pretty much all of them.

## Testing

### Sanity Tests

We recently migrated from Travis to Github Actions. It is a little more obvious in Github Actions now the tests with Ansible Base and Ansible Core for the sanity tests. These were and are in place.

### CI Scheduled Execution

There is now an additional scheduled task to run the CI every Monday morning. This was _added_ over the weekend.

**Reply by tadeboro** [reply]
I did another review, and there are still two issues that we need to resolve.

The release policy is the first and the more important one of the two issues because it directly affects Ansible users. For example, the policy states that collection maintainers will announce deprecations in advance, but this was not the case when removing the data parameter in version 3.0.0.

The second issue is incomplete sanity test coverage of Ansible versions and some failing sanity checks. I described what needs fixes in the relevant comments in the review.

**Reply by jvanderaa** [reply]
So is the wish to have us released a version 2.0.2 with deprecation notices with zero code changes? We had to make the move as we understood what was changed in Ansible so that the modules would work with Ansible Core. The modules would not work otherwise.

If it was not for the broken state with Ansible core we would not have made the sudden change.

**Reply by jvanderaa** [reply]
Thank you for the updated info on the tests. We will get those updated.

**Reply by tadeboro** [reply]
> So is the wish to have us released a version 2.0.2 with deprecation notices with zero code changes? We had to make the move as we understood what was changed in Ansible so that the modules would work with Ansible Core. The modules would not work otherwise.

Well, this is what you promise in your release policy. So yes, I would say that there should be one last version of the collection in the 2.x.y series that deprecates the removed parameter.

> If it was not for the broken state with Ansible core we would not have made the sudden change.

The `netbox.netbox` collection maintainers solved this issue in a backward-compatible way, so removing the `data` parameter was not strictly needed. But there is nothing wrong with your approach, you just need to be a bit more proactive when it comes to informing users about breaking changes.

**Reply by jvanderaa** [reply]
Sounds good, we will rectify and try to figure out how to put that into a release of 2.0.2 if that will satisfy this.

We are heavy _users_ of Ansible ourselves, me more so a user than a developer of the modules. So I get where the stability of the modules are coming from. We are not in the business of just making significant changes lightly. So in hindsight the thing to have done when we made the https://github.com/nautobot/nautobot-ansible/discussions/70 announcement of the upcoming changes, that should have also included a 2.0.2 release with deprecation notices within Ansible itself?

**Reply by tadeboro** [reply]
> We are heavy _users_ of Ansible ourselves, me more so a user than a developer of the modules. So I get where the stability of the modules are coming from. We are not in the business of just making significant changes lightly. So in hindsight the thing to have done when we made the [nautobot/nautobot-ansible#70](https://github.com/nautobot/nautobot-ansible/discussions/70) announcement of the upcoming changes, that should have also included a 2.0.2 release with deprecation notices within Ansible itself?

Yes. This way users of 2.x.y series will be notified that things will change in the near future even if they do not read the announcements on the web.

**Reply by tadeboro** [reply]
I did another round of reviews and the CI setup is still not fulfilling the inclusion requirements. The testing matrix needs to be extended with all supported Ansible versions.

**Reply by jvanderaa** [reply]
@tadeboro it does have all three. [PR #96](https://github.com/nautobot/nautobot-ansible/pull/96) added this for the version that is looking to be included. Version 2 is not for inclusion and does not support Ansible 5 in any way.

**Reply by tadeboro** [reply]
Version 2 of the collection is not a problem anymore (you did all that was required). However, I am having trouble finding a suitable test matrix for the latest version of the code. For example, in https://github.com/nautobot/nautobot-ansible/pull/97/checks, I see three checks that run sanity tests: test (3.7), test (3.8), and test (3.9). You define those at https://github.com/nautobot/nautobot-ansible/blob/2db3228f37aad6708d5f0134acaeb091f1b9b47f/.github/workflows/ci.yml#L24-L41. As far as I can tell, those tests only use `ansible-base-2.10` (https://github.com/nautobot/nautobot-ansible/blob/2db3228f37aad6708d5f0134acaeb091f1b9b47f/poetry.lock#L48-L54).

Or am I missing something here?

**Reply by jvanderaa** [reply]
Ok, that is what I thought, we definitely do not have the matrix defined terribly well, but it is there. 

On https://github.com/nautobot/nautobot-ansible/blob/develop/.github/workflows/ci.yml#L52-L63 we have the lines that define the test matrix of the versions. Base being 2.10, core being 2.11, and legacy being 2.9:

```
        include:
          - python-version: "3.7"
            nautobot-version: "1.0.3"
            ansible-release: "base"
          - python-version: "3.8"
            nautobot-version: "1.1.1"
            ansible-release: "core"
          - python-version: "3.8"
            nautobot-version: "1.1.1"
            ansible-release: "legacy"
          - python-version: "3.8"
            nautobot-version: "1.1.3"
            ansible-release: "core"
```

Then on the lines https://github.com/nautobot/nautobot-ansible/blob/develop/.github/workflows/ci.yml#L72-L89 we do the pip install of each version of Ansible into the containers, modifying the base container version. Not elegant, but does get the testing done.

```
      - name: "Install poetry"
        if: "${{ matrix.ansible-release == 'core' }}"
        run: "pip install poetry"
      - name: "Remove ansible-base"
        if: "${{ matrix.ansible-release == 'core' }}"
        run: "poetry remove ansible-base"
      - name: "Add ansible-core"
        if: "${{ matrix.ansible-release == 'core' }}"
        run: "poetry add ansible-core"
      - name: "Install poetry"
        if: "${{ matrix.ansible-release == 'legacy' }}"
        run: "pip install poetry"
      - name: "Remove ansible-base"
        if: "${{ matrix.ansible-release == 'legacy' }}"
        run: "poetry remove ansible-base"
      - name: "Add Ansible 2.9"
        if: "${{ matrix.ansible-release == 'legacy' }}"
        run: "poetry add ansible@^2.9"
```

**Reply by tadeboro** [reply]
I saw that matrix, but since it was defined on the integration job, I assumed it did not run sanity tests. But this does indeed cover the 2.9, 2.10, and 2.11 versions of ansible(-base|-core), but you are still missing ansible-core-2.12. You will probably need to update the matrix a bit since base, core, and legacy labels are not sufficient to tell apart the ansible-core-2.11 and ansible-core-2.12.

**Reply by jvanderaa** [reply]
Getting the latest beta release of `ansible-core` 2.12 into the pipeline now. I had two updates to make on the sanity for it. I'm just waiting for one approval, which may not be in when you get started. I'll make one more comment at that time.

**Reply by jvanderaa** [reply]
The PR is now merged in that has a test for ansible-core 2.12 Beta 2 included.

**Reply by tadeboro** [reply]
I did my final review and as far as I am concerned, the collection is ready for inclusion in Ansible 5.

**Reply by jvanderaa** [reply]
Thank you @tadeboro for all of the guidance along the way!

**Reply by dmsimard** [reply]
> I did not finish my review since there seem to be some major issues with the licensing in this collection. For example, commit 301e66c8 stripped copyright and licensing information from modules ([nautobot/nautobot-ansible@301e66c#diff-1aaebd78d9857e913ef7e74ca7a6448ee8bf3a04a583666f7882e56b378e98a4](https://github.com/nautobot/nautobot-ansible/commit/301e66c846fda9641b15aa50a96c704f53bad681#diff-1aaebd78d9857e913ef7e74ca7a6448ee8bf3a04a583666f7882e56b378e98a4) is one such module modification). As far as I can tell, this is not legal since it violates the original copyright and license because the new repository seems to be under Apache 2.0 now (again, I am not sure if changing the license here was a valid move).

For posterity, I want to point out that the original licensing and copyright attributions have been restored in https://github.com/nautobot/nautobot-ansible/commit/c34c43a0001dd55c7a1a9409b676feb0dfc76914.

**Comment by tadeboro** [comment]
There are less than two weeks left for collections to get ready for inclusion in Ansible 5 (see [roadmap](https://docs.ansible.com/ansible/devel/roadmap/COLLECTIONS_5.html#release-schedule) for more details). This candidate still has unresolved issues and is not ready for inclusion right now.

**Reply by jvanderaa** [reply]
Thanks @tadeboro, with the latest merge of a couple hours ago I believe we have covered all of the items requested. 

On the 2.x front, we have released a new release in the 2.x release with the appropriate deprecations. There is also a `develop2` [branch](https://github.com/nautobot/nautobot-ansible/tree/develop2) for any future updates to that code.

On the Ansible testing front, [PR #96](https://github.com/nautobot/nautobot-ansible/pull/96) added in the Ansible 2.9 setup as mentioned.

I appreciate the help through this, and including this reminder.

**Comment by jvanderaa** [comment]
@tadeboro were we good on the inclusion?

**Reply by dmsimard** [reply]
@jvanderaa as discussed in #ansible-network, for posterity: I'm sorry we didn't get back to you. We discussed nautobot's inclusion in a [community meeting](https://github.com/ansible/community/issues/539#issuecomment-953223769) but unfortunately we didn't manage to get more eyes on the review in time before the deadline for Ansible 5.

The collection can still be included in Ansible 6 if there is interest and gets at least an additional review.

In the meantime the collection can still be used alongside Ansible 5 but you would need to install it from Galaxy.

**Reply by jvanderaa** [reply]
Let's plan for that inclusion to Ansible 6, if we can't get something into Ansible 5.1 or other dot release.

**Reply by jvanderaa** [reply]
@dmsimard is this all set for Ansible 6 or is there more that is needed yet?

**Comment by jvanderaa** [comment]
@dmsimard @tadeboro are we all set to go for Ansible 6?

**Comment by jvanderaa** [comment]
@gundalow will we be good for the next community version of Ansible to have the modules included?

**Comment by Andersson007** [comment]
# Ansible Collections Checklist (short version)

_For details about the following points, refer to the [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst)._

Every comment should say whether the reviewer expects it to be addressed, or whether it's optional.

**Public availability and communication:**
- [x] published on [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [x] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
- [x] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues
- [x] have a public SCM repository, releases of the collection are tagged in this repository

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
- [x] follows [licensing rules](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#licensing)
- [x] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
  **SHOULD FIX:** https://github.com/nautobot/nautobot-ansible/pull/119
- [x] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html); as well as these other requirements:
  - [x] modules that only gather information are named `<something>_info`
  - [x] modules that return `ansible_facts` are named `<something>_facts` and do not return non-facts
  - [x] other modules must not allow querying information using specific `state` option values, or similar mechanisms (like `state=get` or `state=query`).  These features should be moved to `<something>_info` or `<something>_fact` modules.
  - [x] `check_mode` is supported in all `_info` and `_facts` modules
- [x] supports Python 2.6 or greater and Python 3.5 or greater. If it does not, read the [full guidelines](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#python-compatibility) to see if you qualify for an exception and document the unsupported [Python versions](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#ansible-and-python-3) in the collection ``README.md`` and in every module and plugin (or in doc fragments)
- [x] only uses the [allowed plugin types](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#modules-plugins) in the `plugins/` directory
- [x] has `README.md`
- [x] documentation, examples, and return sections use FQCNs for the `M(..)` [format macros](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation) when referring to modules
- [x] modules (or plugins) from ansible-core use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module or plugin was added
- [x] FQCNs are used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-core

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0, and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
- [x] has changelog, preferably with `changelogs/changelog.yaml`
- [x] collection repository should not contain any large objects (binaries) comparatively to the current Galaxy tarball size limit of 20 MB. For example, package installers for testing purposes shouldn't be added.
- [x] collection repository should not contain any unnecessary files like, for example, temporary files created by development tools. Temporary files MUST be added to `.gitignore`.

**Tests:**

- [x] passed `ansible-test sanity`
  **MUST FIX**: please clarify the question https://github.com/ansible-collections/ansible-inclusion/discussions/28#discussioncomment-2227011
- [x] if `test/sanity/ignore*.txt` exists, it does not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [x] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
- [x] all CI tests run against every pull request
- [x] all CI tests run against a commit that releases the collection; if they don't pass, the collection won't be released
- [x] all CI tests run regularly (nightly, or at least once per week)

**Reply by Andersson007** [reply]
@jvanderaa thanks for submitting the collection! If you fix the sanity, i don't see any other obstacles for inclusion of the collection

**Reply by jvanderaa** [reply]
We will check in on the sanity tests. I know we had some particular setup around getting the docs published appropriately to RTD.

**Comment by Andersson007** [comment]
when running sanity tests locally they stumbled upon the following:
```
Running sanity test "shebang"
ERROR: Found 61 shebang issue(s) which need to be resolved:
ERROR: docs/plugins/aggregate_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/cable_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/circuit_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/circuit_termination_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/circuit_type_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/cluster_group_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/cluster_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/cluster_type_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/console_port_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/console_port_template_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/console_server_port_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/console_server_port_template_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/device_bay_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/device_bay_template_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/device_interface_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/device_interface_template_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/device_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/device_role_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/device_type_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/front_port_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/front_port_template_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/gql_inventory_inventory.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/index.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/inventory_inventory.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/inventory_item_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/ip_address_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/ipam_role_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/lookup_graphql_lookup.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/lookup_lookup.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/manufacturer_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/nautobot_server_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/platform_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/power_feed_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/power_outlet_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/power_outlet_template_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/power_panel_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/power_port_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/power_port_template_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/prefix_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/provider_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/query_graphql_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/rack_group_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/rack_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/rack_role_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/rear_port_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/rear_port_template_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/region_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/rir_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/route_target_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/service_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/site_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/status_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/tag_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/tenant_group_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/tenant_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/virtual_chassis_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/virtual_machine_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/vlan_group_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/vlan_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/vm_interface_module.rst:0:0: file without shebang should not be executable
ERROR: docs/plugins/vrf_module.rst:0:0: file without shebang should not be executable
```
Is it needed that rst files are executable?

**Reply by FragmentedPacket** [reply]
3.3.1 has been released to address this: https://galaxy.ansible.com/networktocode/nautobot

Is this the last remaining item for inclusion?

**Reply by Andersson007** [reply]
@FragmentedPacket cool, i'll update my review and will create a topic for inclusion, thanks!

**Comment by felixfontein** [comment]
While glancing over this collection, I found a couple of issues:

- [x] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
      **SHOULD FIX:** I found multiple instances of:
```
      query_params:
        description:
          - This can be used to override the specified values in ALLOWED_QUERY_PARAMS that is defined
          - in plugins/module_utils/utils.py and provides control to users on what may make
          - an object unique in their environment.
```
Here a single sentence is spread over three paragraphs (!). This should be a single paragraph.

- [x] has changelog, preferably with `changelogs/changelog.yaml`
      **SHOULD FIX:** three instances of `namespace: networktocode.nautobot` in changelogs/changelog.yaml should be changed to `namespace: ''`
- [ ] passed `ansible-test sanity`
      **MUST FIX:** `ansible-test sanity --docker` does not pass with ansible-core 2.12. I suspect it also does not pass with ansible-core 2.11, and it won't pass with ansible-core 2.13 / devel branch either.

**Reply by FragmentedPacket** [reply]
Hi Felix,

I was looking into this today and fixed the changelog namespace issues before I moved onto the `ansible-test sanity` and it appears that it runs the sanity tests perfectly fine. https://github.com/nautobot/nautobot-ansible/runs/5442967949?check_suite_focus=true#step:14:406

Our CI is a bit weird and needs optimized (my fault), but it does run the sanity tests along the different Ansible versions. It doesn't run with the `--docker` flag, but from my understanding, that is if the collection doesn't have dependencies and runs in Ansible's default container, but to my knowledge, it does't run any different tests.

**Reply by felixfontein** [reply]
Hi @FragmentedPacket, you are only running a subset of all tests there since you explicitly limit the number of Python versions used to a single one. Even if you would not specify `--python` you likely limit the number of Python versions used to a subset of the ones needed to run all the tests. You should really run `ansible-test sanity --docker` if you are not an expert on how ansible-test works and can guarantee that all tests are still run.

**Reply by FragmentedPacket** [reply]
We only support Python 3.6 and newer so that's why we limit it via the Python version. I ran `ansible-tests sanity --docker` locally and it only fails on things related to Python version 2.6/2.7/3.5.

**Reply by felixfontein** [reply]
In that case you need to add corresponding entries to the `ignore-2.x.txt` files, respectively use an ansible-test configuration file for new enough ansible-test versions ([template](https://github.com/ansible/ansible/blob/devel/test/lib/ansible_test/config/config.yml)).

**Reply by FragmentedPacket** [reply]
How do we install dependencies into the docker container that these tests run in? We have a third party dependency (`pynautobot`).

**Reply by felixfontein** [reply]
There is no need to install anything in the docker container for `ansible-test sanity`, it brings anything that it needs. The code has to handle `pynautobot` not being present anyway. If you don't do that correctly, the tests are supposed to fail. (See https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/dev_guide/testing/sanity/import.rst on how to handle imports correctly.)

**Reply by Andersson007** [reply]
So let's summarize maybe what the maintainers should do to start the inclusion process. Do they need to add `ignore-2.x.txt` files and that's it?

**Reply by jvanderaa** [reply]
Now that I'm back from leave I will try to get this squared away. Message sent in the Ansible Network Developer slack.

**Reply by Andersson007** [reply]
@jvanderaa that's cool, thanks! We'll be looking forward to your comment that the collection is ready.

**Reply by jvanderaa** [reply]
Now with https://github.com/nautobot/nautobot-ansible/pull/131 merged in, things are back in a good state (2.12 ignore was added which was not needed at the end of last inclusion checks)

**Reply by Andersson007** [reply]
@jvanderaa thanks for the info! I'll be on PTO, will take a look in a week or two

**Reply by Andersson007** [reply]
@jvanderaa i've run sanity tests against `devel`, `stable-2.13`, `stable-2.12`, `stable-2.11` and they've failed in all of the cases.
Could you please fix the collection to make it passing the tests?

**Reply by FragmentedPacket** [reply]
Is there a way for it to only run on the `plugins/` folder?

**Reply by kristianheljas** [reply]
Although [the docs](https://docs.ansible.com/ansible/latest/dev_guide/testing_sanity.html#how-to-run) do not state it explicitly, you can specify a folder as well: `ansible-test sanity plugins/`

I will create a PR to the docs.

**Reply by Andersson007** [reply]
as @felixfontein spotted
>it has to pass for the whole collection (as published on galaxy), otherwise we cannot automate sanity tests

**Reply by felixfontein** [reply]
And another one: by restricting it to `plugins/`, you are missing validation of changelog fragments (if you use them), of meta/runtime.yml, and potentially other things.
